hudi-common/src/main/java/org/apache/hudi/common/table/log/block/HoodieAvroDataBlock.java

```diff
@@ -224,6 +249,159 @@ public IndexedRecord next() {
+      return new StreamingRecordIterator(dataBlock.readerSchema, dataBlock.getSchemaFromHeader(), inputStream, contentLocation, bufferSize);
+    }
+
+    @Override
+    public void close() {
+      this.decoderCache.remove();
+      this.buffer = null;
+      try {
+        this.inputStream.close();
+      } catch (IOException ex) {
+        throw new HoodieIOException("Failed to close input stream", ex);
+      }
+    }
+
+    @Override
+    public boolean hasNext() {
+      return readRecords < totalRecords;
+    }
+
+    @Override
+    public IndexedRecord next() {
+      try {
+        ensureBufferHasData(RECORD_LENGTH_BYTES);
+
+        // Read the record length
+        int recordLength = buffer.getInt();
+
+        // Ensure buffer is large enough and has enough data
+        ensureBufferHasData(recordLength);
+
+        // Decode the record
+        BinaryDecoder decoder = DecoderFactory.get().binaryDecoder(buffer.array(), buffer.position(), recordLength, this.decoderCache.get());
+        this.decoderCache.set(decoder);
+        IndexedRecord record = this.reader.read(null, decoder);
+        buffer.position(buffer.position() + recordLength);
+        this.readRecords++;
+        if (this.promotedSchema.isPresent()) {
+          return HoodieAvroUtils.rewriteRecordWithNewSchema(record, this.promotedSchema.get());
+        }
+        return record;
+      } catch (IOException e) {
+        throw new HoodieIOException("Unable to convert bytes to record", e);
+      }
+    }
+
+    /**
+     * Ensures that the buffer contains at least the specified amount of data.
+     *
+     * This method checks if the buffer has the required amount of data. If not, it attempts to fill the buffer
+     * by reading more data from the input stream. If the buffer's capacity is insufficient, it allocates a larger buffer.
+     * If the end of the input stream is reached before the required amount of data is available, an exception is thrown.
+     *
+     * @param dataLength the amount of data (in bytes) that must be available in the buffer.
+     * @throws IOException if an I/O error occurs while reading from the input stream.
+     * @throws HoodieException if the end of the input stream is reached before the required amount of data is available.
+     */
+    private void ensureBufferHasData(int dataLength) throws IOException {
+      // Check if the current buffer has enough space to read the required data length
+      if (buffer.capacity() - buffer.position() < dataLength) {
+        buffer.compact();
+        // Reset the buffer to read mode
+        buffer.flip();
+      }
+
+      // Check again if the buffer still doesn't have enough space after compaction
+      if (buffer.capacity() - buffer.position() < dataLength) {
+        ByteBuffer newBuffer = ByteBuffer.allocate(buffer.position() + dataLength);
+        newBuffer.put(buffer);
+        // Reset the new buffer to read mode
+        newBuffer.flip();
```

Comment:
should we do `flip` on the fresh new buffer, because the following code has the check `buffer.remaining() < dataLength`, but the limit has been changed after the flip operation.

Reply:
> should we do `flip` on the fresh new buffer, because the following code has the check `buffer.remaining() < dataLength`, but the limit has been changed after the flip operation.

Hello Danny, yes, after calling `flip()` in write mode, the `limit()` will equal the size of the readable data, indicating that you can only read up to this position. This ensures that subsequent read operations can function correctly.

Why do we always call `flip()` after writing? Because once writing is complete, we need to switch back to read mode.